lib/active_shipping/shipment_packer.rb

```diff
@@ -11,63 +11,90 @@ class ExcessPackageQuantity < StandardError; end
+
+        state = :package_empty
+        while state != :packing_finished
+          case state
+          when :package_empty
+            package_weight, package_value = 0, 0
+            state = :filling_package
+          when :filling_package
+            items.each do |item|
+              quantity = determine_fillable_quantity_for_package(item, maximum_weight, package_weight)
+              package_weight += item_weight(quantity, item[:grams])
+              package_value += item_value(quantity, item[:price])
+              item[:quantity] = item[:quantity].to_i - quantity
+            end
+
+            items.reject! { |i| i[:quantity].to_i == 0 }
+            state = :package_full
+          when :package_full
+            packages << ActiveShipping::Package.new(package_weight, dimensions, :value => package_value, :currency => currency)
+            state = items.any? ? :package_empty : :packing_finished
+          end
         end
+
+        packages
       end
 
-      items = items.map(&:dup).sort_by! { |i| i[:grams].to_i }
-
-      state = :package_empty
-      while state != :packing_finished
-        case state
-        when :package_empty
-          package_weight, package_value = 0, 0
-          state = :filling_package
-        when :filling_package
-          items.each do |item|
-            quantity = if item[:grams].to_i <= 0
-              item[:quantity].to_i
-            else
-              # Grab the max amount of this item we can fit into this package
-              # Or, if there are fewer than the max for this item, put
-              # what is left into this package
-              [(maximum_weight - package_weight) / item[:grams].to_i, item[:quantity].to_i].min
-            end
+      private
 
-            item_weight = quantity * item[:grams].to_i
-            item_value = quantity * Package.cents_from(item[:price])
+      def calculate_total_weight(items, maximum_weight)
+        items.inject(0) do |total_weight, item|
+          total_weight += item[:quantity].to_i * item[:grams].to_i
 
-            package_weight += item_weight
-            package_value += item_value
+          if overweight_item?(item[:grams], maximum_weight)
+            raise OverweightItem, "The item with weight of #{item[:grams]}g is heavier than the allowable package weight of #{maximum_weight}g"
+          end
 
-            item[:quantity] = item[:quantity].to_i - quantity
+          if excess_package_quantity?(total_weight, maximum_weight)
+            raise ExcessPackageQuantity, "Unable to pack more than #{EXCESS_PACKAGE_QUANTITY_THRESHOLD} packages"
           end
 
-          items.reject! { |i| i[:quantity].to_i == 0 }
+          total_weight
+        end
+      end
+
+      def overweight_item?(grams, maximum_weight)
+        grams.to_i > maximum_weight
+      end
+
+      def excess_package_quantity?(total_weight, maximum_weight)
+        total_weight > (maximum_weight * EXCESS_PACKAGE_QUANTITY_THRESHOLD)
```

Comment:
Throwing a question out there: does this make sense?

The error is saying "Unable to pack more than 10000 packages" but the validation is checking if `total_weight > maximum_weight * 10000`. Which one are we validating; # of packages or total weight?